### PR TITLE
Have Legacy Mode use a Popup Window

### PR DESCRIPTION
**Pull request by jisqyv**
Starting with Chrome 94, we can no longer use Ajax requests to
communicate via http with the Companion. We also cannot communicate
with http if we are in a secure context.

This change introduces a Popup window which is loaded from the
Companion itself and can therefore communicate with the Companion. Our
main window (from replmgr.js) will now communicate with this Popup
which will act as a go between. This not only permits Chrome 94+ to
work but also permits us to load App Inventor in a secure context.

This code is both forward and backward compatible. An older Companion
will use Legacy Mode as it always has (and be subject to the
restrictions involved). Similarly the new Companion will work with an
older version of App Inventor.

The new Companion indicates that it can use a popup by sending the
"useproxy" flag to the Rendezvous server which is then queried for in
replmgr.js

Change-Id: I43005b78c190dfb8446308f692dc39dec51befb1

### CODE REVIEW

**Comment by jisqyv** [comment]
Note: There is still more work to be done here (including rebasing to ucr). In particular, I need to gracefully handle the case where the popup is blocked. I am also experimenting with how old a browser can run the Popup code, given it is using newer Javascript and DOM features (aka async/await and fetch). If necessary I will transpile the code and include the actual source somewhere else in the source hierarchy, perhaps under misc along with the webpack config and package.json to build it (I am currently experimenting with npm/webpack/babel).

**Comment by halatmit** [comment]
Nice to see that you have a solution .  Is there some way we can add a
message that will encourage people not to use an
old companion?
== Hal

Hal Abelson
***@***.***
Prof. of Comp. Sci. and Eng.
MIT Dept. of Elec. Eng. and Comp. Sci.

**Comment by jisqyv** [comment]
Testing: The code as is (not transpiled) works at least back to Chrome 50 (4/2016) and Firefox 52 (3/2017)

**Comment by jisqyv** [comment]
@halatmit Most people have their Companion auto-updated from the Play Store.

We released version 2.62 on 11/9 and there are already 98,000 unique copies as seen by the Rendezvous server while the previous version (2.61) has 73,000 unique copies as soon by the Rendezvous server since 11/9. Older versions are much fewer.

**Comment by halatmit** [comment]
Good that the autoupdating it doing well.

== Hal

Hal Abelson
***@***.***
Prof. of Comp. Sci. and Eng.
MIT Dept. of Elec. Eng. and Comp. Sci.

**Comment by jisqyv** [comment]
Latest commit message isn't quite right, but the code is good.

File: appinventor/blocklyeditor/src/replmgr.js
Comment (ellelili2025) [inline appinventor/blocklyeditor/src/replmgr.js:296]: Is this supposed to be rs.phoneState.initialized = false ?

Comment (jisqyv) [reply appinventor/blocklyeditor/src/replmgr.js]: Fixed

**Comment by jisqyv** [comment]
Note: I have converted this back to a draft. It is now based on #2664 so should not be merged until after that PR is merged and this one appropriately tested and rebased.

**Comment by jisqyv** [comment]
I've just converted this PR to a DRAFT. I'm going to refactor it into two commits. One for master, with just the website side changes, and the other for ucr with the Companion changes.

File: appinventor/blocklyeditor/src/replmgr.js
Comment (ewpatton) [inline appinventor/blocklyeditor/src/replmgr.js:1830]: Maybe document what happens if we fail.

**Review by ewpatton — CHANGES_REQUESTED** [review]
A few minor comments for us to think about.

File: appinventor/blocklyeditor/src/replmgr.js
Comment (jisqyv) [reply appinventor/blocklyeditor/src/replmgr.js:1819]: arrayview is a Uint8Array "view" of the arraybuf, which is used on line 1829

File: appinventor/blocklyeditor/src/replmgr.js
Comment (jisqyv) [reply appinventor/blocklyeditor/src/replmgr.js:1830]: Actually, I'm not sure what happens. postMessage() isn't documented to return an error, though other sources reference that it can throw an exception. However, I don't think it ever will given the way we use it (no use of shared memory).

File: appinventor/blocklyeditor/src/replmgr.js
Comment (jisqyv) [reply appinventor/blocklyeditor/src/replmgr.js:1670]: This is not an error situation. This is the promise sending a reject. This happens when we put up the companion out of date error and the user presses OK (i.e., do not continue). In this case, within the promise the connection attempt is aborted, and then it does the reject, which causes the whole process to stop.

File: appinventor/components/src/com/google/appinventor/components/runtime/util/AppInvHTTPD.java
Comment (jisqyv) [reply appinventor/components/src/com/google/appinventor/components/runtime/util/AppInvHTTPD.java:266]: I would rather not add this header on this round. Before we add this header, I would want to test it on a broad range of browsers (actually browser versions).

**Review by ewpatton — APPROVED** [review]

**Comment by ewpatton** [comment]
I'm approving this for now, but we should definitely keep an eye on whether the header will be needed. Ideally it would be nice if we don't have to scramble.

File: appinventor/blocklyeditor/src/msg/ai_blockly/messages.json
Comment (jisqyv) [reply appinventor/blocklyeditor/src/msg/ai_blockly/messages.json:767]: I may put a "Why do I have to do this" link in the dialog that will link to a page on the website. Unfortunately, exactly how you unblock pop-ups is both browser and extension dependent, and some schools may block such unblocking!

**Pull request merged**
